### PR TITLE
Finite State Machine Play Architecture

### LINKED ISSUE #2382: Set up the Play FSM Architecture
### Description of the task
* Set up the play fsm architecture according to https://docs.google.com/document/d/1ZWv_hviRrp8yjL-gb8m4BwL3Y20nQlJ9fF_xyzGpmFc/edit?usp=sharing
* Set up some examples

### Acceptance criteria

- [ ] as described

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
* Set up the infrastructure for the new [play architecture](https://docs.google.com/document/d/1ZWv_hviRrp8yjL-gb8m4BwL3Y20nQlJ9fF_xyzGpmFc/edit?usp=sharing)
* Renamed shoot or pass play to offense play
* Created shoot or pass play fsm to handle the offensive aspects of the old shoot or pass play
* Created a simple crease defense play fsm for crease defense with up to 3 robots
* Integrated the shoot or pass play fsm and crease defense play fsm into the offense play fsm

### Testing Done
fsm tests for shoot or pass play fsm and crease defense play fsm
simulated play tests for offense play fsm

### Resolved Issues
resolves #2382

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/ai/hl/stp/play/crease_defense/crease_defense_play_fsm.h [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        // number of additional crease defenders on top of the minimum of 1
+        unsigned int num_additional_crease_defenders_tactics;
+        // The maximum allowed speed mode
+        MaxAllowedSpeedMode max_allowed_speed_mode;
+    };
+
+    DEFINE_PLAY_UPDATE_STRUCT_WITH_CONTROL_AND_COMMON_PARAMS
+
+    /**
+     * Creates a crease defense play fsm
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/crease_defense/crease_defense_play_fsm.h]: super nit: FSM or fsm?

File: src/software/ai/hl/stp/play/crease_defense/crease_defense_play_fsm.h [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+     *
+     * @param play_config the play config for this play fsm
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/crease_defense/crease_defense_play_fsm.h]: super nit: FSM or fsm?

File: src/software/ai/hl/stp/play/crease_defense/crease_defense_play_fsm.cpp [hunk 3]
```diff
@@ -0,0 +1,42 @@
+#include "software/ai/hl/stp/play/crease_defense/crease_defense_play_fsm.h"
+
+void CreaseDefensePlayFSM::defendDefenseArea(const Update& event)
+{
+    unsigned int num_defenders = event.common.num_tactics;
+    // TODO (#2385): Handle more than three defenders
+    if (num_defenders > 3)
```
Comment (akhilveeraghanta) [inline src/software/ai/hl/stp/play/crease_defense/crease_defense_play_fsm.cpp:12]: do all plays need to have checks like this?

File: src/software/ai/hl/stp/play/crease_defense/crease_defense_play_fsm.h [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+     */
+    explicit CreaseDefensePlayFSM(std::shared_ptr<const PlayConfig> play_config)
+        : play_config(play_config), crease_defenders({})
+    {
+    }
+
+    /**
+     * Action to defend the defense defense area
```
Comment (akhilveeraghanta) [inline src/software/ai/hl/stp/play/crease_defense/crease_defense_play_fsm.h]: nit: action to defend the defense defense

File: src/software/ai/hl/stp/play/offense/offense_play_fsm.cpp [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+{
+}
+
+void OffensePlayFSM::updateOffense(const Update& event)
+{
+    PriorityTacticVector tactics_to_return;
+    unsigned int num_shoot_or_pass =
+        event.common.num_tactics > 3 ? event.common.num_tactics - 2 : 1;
+    unsigned int num_defenders =
+        event.common.num_tactics > 3 ? 2 : event.common.num_tactics - 1;
```
Comment (akhilveeraghanta) [inline src/software/ai/hl/stp/play/offense/offense_play_fsm.cpp]: ternary operators :eyes: (we forbid these in our style guide)

File: src/software/ai/hl/stp/play/offense/offense_play_fsm.h [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+    /**
+     * Updates the offense play
+     *
+     * @param event the OffensePlayFSM event
+     */
+    void updateOffense(const Update& event);
+
+    /**
+     * Guard for whehter offense is done
```
Comment (akhilveeraghanta) [inline src/software/ai/hl/stp/play/offense/offense_play_fsm.h]: whehter -> whether

**Review by akhilveeraghanta — COMMENTED** [review]
looks awesome! couple minor nits

File: src/software/ai/hl/stp/play/crease_defense/crease_defense_play_fsm.cpp (hunk 3 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/play/crease_defense/crease_defense_play_fsm.cpp:12]: no, all plays should handle any number of robots (hopefully) and #2385 should make that so for the crease defense play

File: src/software/ai/hl/stp/play/offense/offense_play_fsm.cpp (hunk 5 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/play/offense/offense_play_fsm.cpp]: you got me

File: src/software/ai/hl/stp/play/crease_defense/crease_defense_play_fsm_test.cpp [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        PlayUpdate(world, 3, [](PriorityTacticVector new_tactics) {})));
+
+    EXPECT_TRUE(fsm.is(boost::sml::X));
+
+    // Change the number of defenders
+    fsm.process_event(CreaseDefensePlayFSM::Update(
+        CreaseDefensePlayFSM::ControlParams{
+            .enemy_threat_origin    = Point(),
+            .max_allowed_speed_mode = MaxAllowedSpeedMode::PHYSICAL_LIMIT},
+        PlayUpdate(world, 3, [](PriorityTacticVector new_tactics) {})));
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/crease_defense/crease_defense_play_fsm_test.cpp]: Are the number of defenders being changed here? Looks like it's still 3

File: src/software/ai/hl/stp/play/shoot_or_pass/shoot_or_pass_play_fsm.cpp [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+                                                   ->getPassScoreRampDownDuration()
+                                                   ->value();
+        pass_eval = pass_generator.generatePassEvaluation(event.common.world);
+        best_pass_and_score_so_far = pass_eval.getBestPassOnField();
+
+        // update the best pass in the attacker tactic
+        attacker_tactic->updateControlParams(best_pass_and_score_so_far.pass, false);
+
+        // If we've assigned a robot as the passer in the PassGenerator, we
+        // lower our threshold based on how long the PassGenerator as been
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/shoot_or_pass/shoot_or_pass_play_fsm.cpp]: nit: how long the PassGenerator as been running -> how long the PassGenerator has been running

File: src/software/ai/hl/stp/play/shoot_or_pass/shoot_or_pass_play_fsm.h [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+        DEFINE_SML_ACTION(takePass)
+
+        DEFINE_SML_GUARD(passFound)
+        DEFINE_SML_GUARD(shouldAbortPass)
+        DEFINE_SML_GUARD(passCompleted)
+        DEFINE_SML_GUARD(tookShot)
+
+        return make_transition_table(
+            // src_state + event [guard] / action = dest_state
+            *StartState_S + Update_E / startLookingForPass_A        = AttemptShotState_S,
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/shoot_or_pass/shoot_or_pass_play_fsm.h:126]: startLookingForPass and lookForPass seem pretty similar, can we combine it or is there a reason it's separate?

**Review by itsarune — COMMENTED** [review]
plays and subplays look insane! I've also got some comments and questions

File: src/software/ai/hl/stp/play/offense/offense_play_fsm.cpp [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+            .max_allowed_speed_mode = MaxAllowedSpeedMode::PHYSICAL_LIMIT},
+        PlayUpdate(event.common.world, num_defenders,
+                   [&tactics_to_return](PriorityTacticVector new_tactics) {
+                       for (const auto& tactic_vector : new_tactics)
+                       {
+                           tactics_to_return.push_back(tactic_vector);
+                       }
+                   })));
+
+    event.common.set_tactics(tactics_to_return);
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/offense/offense_play_fsm.cpp]: did you figure out a way to enforce that we'll always return the correct number of robots?
if not, maybe we can just add a debug warning if a play returns less than the required number of robots in stp for debugging

**Comment by SpeedyPetey** [comment]
Can't really say i know completely what is going on, but looks good to me

File: src/software/ai/hl/stp/play/shoot_or_pass/shoot_or_pass_play_fsm.h (hunk 9 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/play/shoot_or_pass/shoot_or_pass_play_fsm.h:126]: it needs to reset the Attacker and Receiver Tactics when it starts looking for a new pass

File: src/software/ai/hl/stp/play/offense/offense_play_fsm.cpp (hunk 10 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/play/offense/offense_play_fsm.cpp]: I was thinking that we eventually do exhaustive testing for those cases with some simulated testing. In any case, that's out of scope for this PR

File: src/software/ai/hl/stp/play/shoot_or_pass/shoot_or_pass_play_fsm.cpp (hunk 8 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/play/shoot_or_pass/shoot_or_pass_play_fsm.cpp]: fixed

File: src/software/ai/hl/stp/play/crease_defense/crease_defense_play_fsm.h (hunk 2 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/play/crease_defense/crease_defense_play_fsm.h]: fixed

File: src/software/ai/hl/stp/play/crease_defense/crease_defense_play_fsm_test.cpp (hunk 7 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/play/crease_defense/crease_defense_play_fsm_test.cpp]: fixed

File: src/software/ai/hl/stp/play/offense/offense_play_fsm.h (hunk 6 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/play/offense/offense_play_fsm.h]: fixed

File: src/software/ai/hl/stp/play/crease_defense/crease_defense_play_fsm.h (hunk 4 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/play/crease_defense/crease_defense_play_fsm.h]: fixed

File: src/software/ai/hl/stp/play/offense/offense_play_fsm.cpp (hunk 5 above)
Comment (akhilveeraghanta) [reply src/software/ai/hl/stp/play/offense/offense_play_fsm.cpp]: didn't see a change here? seems convenient though so im willing to :see_no_evil:

Comment (jonathanlew) [reply src/software/ai/hl/stp/play/offense/offense_play_fsm.cpp]: I don't think you got the latest changes

**Review by akhilveeraghanta — DISMISSED** [review]
💯

**Review by akhilveeraghanta — DISMISSED** [review]
lgtm!

**Review by itsarune — APPROVED** [review]
lgtm!

**Review by akhilveeraghanta — APPROVED** [review]

**Pull request merged**